include/deal.II/lac/solver_fire.h

```diff
@@ -0,0 +1,392 @@
+ * Objective function:              $E(U)$.     <BR>
+ * Rate of change of unknowns:      $V$.        <BR>
+ * Gradient of the objective
+ * function w.r.t unknowns:         $G$.        <BR>
+ * Mass matrix:                     $M$.        <BR>
+ * Initial guess of unknowns:       $U_0$.      <BR>
+ * Time step:                       $\Delta t$. <BR>
+ *
+ * Given initial values for $\Delta t$, $\alpha = \alpha_0$, $U = U_0$ and $V=0$
+ * along with a given mass matrix $M$,
+ * FIRE algorithm is as follows,
+ * 1. Calculate $G = \nabla E(U)$ and check for convergence.
+ *    Calculate $U$ and $V$ using any common Molecular Dynamics time integrator.
+ *    Here we use simple Euler integration step, <BR>
+ *        $U = U + \Delta t V$,                  <BR>
+ *        $V = V + \Delta t M^{-1} \cdot G$.
+ * 2. Calculate $P = G \cdot V$.
+ * 3. Set $V = (1-\alpha) V + \alpha \frac{|V|}{|G|} G$.
+ * 4. If $P \leq 0$ and number of steps since P was non-negative is larger than
+ *    certain value, then increase time step $\Delta t$ and decrease $\alpha$.
+ * 5. If $P>0$, then decrease the time step, freeze the system i.e., V = 0 and
+ *    reset $\alpha = \alpha_0$.
+ * 6. Return to 1.
+ *
+ * Alse see
+ * <a href="http://onlinelibrary.wiley.com/doi/10.1002/pamm.201110246/full">
+ * Energy-Minimization in Atomic-to-Continuum Scale-Bridging Methods </a> by
+ * Eidel et al. 2011.
+ *
+ * @author Vishal Boddu 2017
+ */
+template<typename VectorType = Vector<double> >
+class SolverFIRE : public Solver<VectorType>
+{
+
+public:
+
+  struct AdditionalData
+  {
+    explicit
+    AdditionalData (const double  initial_timestep    = 0.1,
+                    const double  maximum_timestep    = 1,
+                    const double  maximum_linfty_norm = 1);
+
+    /**
+     * Initial time step.
+     */
+    const double initial_timestep;
+
+    /**
+     * Maximum time step.
+     */
+    const double maximum_timestep;
+
+    /**
+     * Maximum change allowed in any degree of freedom.
+     */
+    const double maximum_linfty_norm;
+
+  };
+
+  /**
+   * Constructor.
+   */
+  SolverFIRE (SolverControl            &solver_control,
+              VectorMemory<VectorType> &vector_memory,
+              const AdditionalData     &data          );
+
+  /**
+   * Constructor. Use an object of type GrowingVectorMemory as a default to
+   * allocate memory.
+   */
+  SolverFIRE (SolverControl         &solver_control,
+              const AdditionalData  &data          );
+
+  /**
+   * Virtual destructor.
+   */
+  virtual ~SolverFIRE();
+
+  /**
+   * Obtain a set of #p u (variables) that minimize an objective function
+   * described by the polymorphic function wrapper @p compute, with a given
+   * preconditioner @p inverse_masses and initial @p u values.
+   * The function @p compute takes in the state of the (u) variables as
+   * argument and returns a pair of objective function's value and
+   * objective function's gradient (with respect to the variables).
```

Comment:
It's not clear to me how it returns these two arguments. Is the gradient what's returned in the first argument of the function object?

With the exception of `vmult` functions we typically try to put input arguments first and output arguments last. Your `compute` function object appears to be the other way around.